File: Recursion_and_Backtracking_pepcoding_level_1/3.Recustion_on_array_list/2.kpc.js

const codes = [
    ".;",
    "abc",
    "def",
    "ghi",
    "jkl",
    "mno",
    "pqrs",
    "tu",
    "vwx",
    "yz"
];

// curr 
const kpce = (str) => {
    if (str.length === 0) return [""];

    const curr = parseInt(str[0]);

    const ans = [];
    for (let code of codes[curr]) {
        ans.push(code);
    }

    return ans;
};

console.log(kpce("2"));

///////////////////////////////////////////
// curr + child
const kpc = (str) => {
    if (str.length === 0) return [""];

    const curr = parseInt(str[0]);

    const childs = kpc(str.slice(1));

    const ans = [];
    for (let code of codes[curr]) {
        for (let child of childs) {
            ans.push(code + child);
        }
    }

    return ans;
};

console.log(kpc(""));

File: Recursion_and_Backtracking_pepcoding_level_1/3.Recustion_on_array_list/2.kpc.rb


def kpc(str, codes)
  return [""] if str.empty?

  single_num = str[0].to_i
  rest = kpc(str[1..-1], codes)

  ans = []
  codes[single_num].chars.each do |code|
    rest.each do |child|
      ans << code + child
    end
  end

  ans
end

codes = [
  ".;",
  "abc",
  "def",
  "ghi",
  "jkl",
  "mno",
  "pqrs",
  "tu",
  "vwx",
  "yz"
]

puts kpc("26", codes).inspect
